File: app/src/main/java/com/example/myapplication/Adapters/ExploreAdapter.java
package com.example.myapplication.Adapters;

import android.content.Context;
import android.content.Intent;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.*;
import com.example.myapplication.Interface.ExploreFragmentContext;
import com.example.myapplication.Interface.MainActivityContext;
import com.example.myapplication.Models.VideoModel;
import com.example.myapplication.R;
import com.example.myapplication.Utils.HorizontalItemDecoration;
import com.example.myapplication.Utils.VideosItemCallback;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public class ExploreAdapter extends RecyclerView.Adapter<ExploreAdapter.ExploreViewHolder> {

    private Context context;
    private Map<String, List<VideoModel>> data;
    private List<String> keys;
    private MainActivityContext mainActivityContext;
    private final VideosItemCallback videosItemCallback = VideosItemCallback.newInstance();
    private static Map<String, Parcelable> recycler_states = null;
    private ExploreFragmentContext exploreFragmentContext;

    public ExploreAdapter(Map<String, List<VideoModel>> data, Context context, MainActivityContext mainActivityContext, ExploreFragmentContext exploreFragmentContext) {
        if(recycler_states == null){
            recycler_states = new HashMap<>();
        }
        this.data = data;
        this.context = context;
        keys = new ArrayList<>(data.keySet());
        this.mainActivityContext = mainActivityContext;
        this.exploreFragmentContext = exploreFragmentContext;
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull @NotNull ExploreViewHolder holder) {
        super.onViewDetachedFromWindow(holder);

        LinearLayoutManager linearLayoutManager = (LinearLayoutManager) holder.recycler_view.getLayoutManager();
        if (linearLayoutManager != null) {
            recycler_states.put(holder.recycler_header.getText().toString(), linearLayoutManager.onSaveInstanceState());
        }
    }

    public void resetRecyclerStates(){
        recycler_states = null;
    }

    @Override
    public void onViewAttachedToWindow(@NonNull @NotNull ExploreViewHolder holder) {
        super.onViewAttachedToWindow(holder);
        LinearLayoutManager linearLayoutManager = (LinearLayoutManager) holder.recycler_view.getLayoutManager();
        if (linearLayoutManager != null) {
            linearLayoutManager.onRestoreInstanceState(recycler_states.get(holder.recycler_header.getText().toString()));
        }
    }

    @NonNull
    @NotNull
    @Override
    public ExploreViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.explore_recycler_item, parent, false);
        RecyclerView recyclerView = view.findViewById(R.id.explore_recycler_item_recycler_view);
        SnapHelper snapHelper = new PagerSnapHelper();
        snapHelper.attachToRecyclerView(recyclerView);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false);
        recyclerView.setLayoutManager(linearLayoutManager);

        HorizontalItemDecoration horizontalItemDecoration = new HorizontalItemDecoration();
        recyclerView.addItemDecoration(horizontalItemDecoration);
        return new ExploreViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull ExploreViewHolder holder, int position) {
        holder.recycler_header.setText(keys.get(position));

        ExploreItemAdapter adapter = new ExploreItemAdapter(videosItemCallback, context, mainActivityContext);
        adapter.submitList(data.get(keys.get(position)));
        holder.recycler_view.setAdapter(adapter);

        holder.imageButton.setOnClickListener(v ->{
            exploreFragmentContext.navigateToVideoListActivity(keys.get(position));
        });
    }

    @Override
    public void onViewRecycled(@NonNull @NotNull ExploreViewHolder holder) {
        super.onViewRecycled(holder);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public static class ExploreViewHolder extends RecyclerView.ViewHolder {

        private TextView recycler_header;
        private RecyclerView recycler_view;
        private ImageButton imageButton;

        public ExploreViewHolder(@NonNull @NotNull View itemView) {
            super(itemView);

            recycler_header = itemView.findViewById(R.id.explore_recycler_item_text_view);
            recycler_view = itemView.findViewById(R.id.explore_recycler_item_recycler_view);
            imageButton = itemView.findViewById(R.id.explore_recycler_more_btn);
        }
    }
}

File: app/src/main/java/com/example/myapplication/Adapters/ExploreAdapter.kt
package com.example.myapplication.Adapters;

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.BottomNavigationFragments.ExploreFragment
import com.example.myapplication.Models.ExploreResponseBody
import com.example.myapplication.Paging.CommonListPagingListeners
import com.example.myapplication.Paging.CommonListPagingViewHolder
import com.example.myapplication.R
import com.example.myapplication.Utils.CustomItemDecoration
import com.example.myapplication.Utils.IntegerToDp
import com.example.myapplication.Utils.VideosItemCallback
import com.example.myapplication.viewmodel.NetworkState
import com.google.android.material.button.MaterialButton

typealias navigateToDestination = (String) -> Unit

class ExploreAdapter(private val exploreFragment: ExploreFragment, private val navigateToMore: navigateToDestination, private val navigateToVideo: navigateToDestination, itemCallback: DiffUtil.ItemCallback<ExploreResponseBody>) : ListAdapter<ExploreResponseBody, ExploreAdapter.ExploreViewHolder>(itemCallback), CommonListPagingListeners {
    private val videosItemCallback = VideosItemCallback.newInstance()
    private var networkState: NetworkState? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ExploreViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.explore_recycler_item, parent, false)
        return ExploreViewHolder(view, this)
    }

    fun onNetworkStateChanged(netState: NetworkState) {
        this.networkState = netState
    }

    override fun onBindViewHolder(holder: ExploreViewHolder, position: Int) {
        val exploreResponseBody = getItem(position)

        if (exploreResponseBody == null) {
            holder.linearLayout.visibility = View.GONE
            holder.handleNetworkStateChanged(networkState!!)
            return
        }
        holder.linearLayout.visibility = View.VISIBLE
        holder.loading_container.visibility = View.GONE
        holder.recycler_header.text = exploreResponseBody.name
        val adapter = ExploreItemAdapter(videosItemCallback, exploreFragment.requireContext(), navigateToVideo)
        holder.recycler_view.adapter = adapter
        adapter.submitList(exploreResponseBody.videos)
        holder.bindClickListener(exploreResponseBody.id)

        if (exploreResponseBody.videos.size > 3) {
            holder.more_btn.visibility = View.VISIBLE
        } else {
            holder.more_btn.visibility = View.GONE
        }
    }

    inner class ExploreViewHolder(itemView: View, adapter: ExploreAdapter) : CommonListPagingViewHolder(itemView, adapter) {

        val recycler_header: TextView = itemView.findViewById(R.id.explore_recycler_item_text_view)
        val recycler_view: RecyclerView = itemView.findViewById(R.id.explore_recycler_item_recycler_view)
        val more_btn: MaterialButton = itemView.findViewById(R.id.explore_recycler_more_btn)
        val linearLayout: LinearLayout = itemView.findViewById(R.id.view_item);

        init {

            val linearLayoutManager = LinearLayoutManager(exploreFragment.requireContext(), LinearLayoutManager.VERTICAL, false)
            recycler_view.layoutManager = linearLayoutManager
            recycler_view.addItemDecoration(CustomItemDecoration(IntegerToDp.intToDp(10), IntegerToDp.intToDp(15), IntegerToDp.intToDp(5), IntegerToDp.intToDp(15)))
        }

        fun bindClickListener(key: String) {
            more_btn.setOnClickListener { navigateToMore(key) }
        }
    }

    override fun retry() {
        exploreFragment.retry()
    }
}
